source/containers/ArchivesList.js

```diff
@@ -20,16 +32,83 @@ function openArchive(dispatch, getState, sourceID) {
     const targetSource = archivesList.find(source => source.id === sourceID);
     // Select source
     dispatch(setSelectedSource(sourceID));
+    const isReadOnly = getSourceReadonlyStatus(sourceID);
+    dispatch(markCurrentSourceReadOnly(isReadOnly));
     // populate groups
     updateCurrentArchive();
     // navigate to archive contents
     dispatch(navigateToGroups({ groupID: "0", title: `🗂 ${targetSource.name}` }));
 }
 
+function performOfflineProcedure(dispatch, getState, sourceID, password, isOffline = false) {
+    return checkSourceHasOfflineCopy(sourceID).then(hasOffline => {
+        if (!hasOffline) {
+            return false;
+        }
+        Alert.alert(
+            isOffline ? "Offline Content (currently offline)" : "Offline Content",
+            "Would you like to try and load this archive in offline (read-only) mode?",
+            [
+                {
+                    text: "Cancel",
+                    style: "cancel",
+                    onPress: () => {
+                        dispatch(setBusyState(null));
+                    }
+                },
+                {
+                    text: "Use Offline",
+                    style: "default",
+                    onPress: () => {
+                        performSourceUnlock(dispatch, getState, sourceID, password, true)
+                            .then(() => {
+                                setTimeout(() => {
+                                    executeNotification(
+                                        "info",
+                                        "Read-Only Mode",
+                                        "This archive was opened in read-only mode due to being offline. " +
+                                            "Changes will not be possible and certain features will be disabled."
+                                    );
+                                }, 1000);
+                            })
+                            .catch(err => {
+                                handleError("Failed unlocking archive", err);
+                            });
+                    }
+                }
+            ]
+        );
+        return true;
+    });
+}
+
+function performSourceUnlock(dispatch, getState, sourceID, password, useOffline = false) {
```

Comment:
also this definition can be changed to redux thunk style: `fn = args => (dispatch, getState) => { ... }`